File: modules/org.openbravo.retail.posterminal/src/org/openbravo/retail/posterminal/master/Product.java
/*
 ************************************************************************************
 * Copyright (C) 2012 Openbravo S.L.U.
 * Licensed under the Openbravo Commercial License version 1.0
 * You may obtain a copy of the License at http://www.openbravo.com/legal/obcl.html
 * or in the legal folder of this module distribution.
 ************************************************************************************
 */
package org.openbravo.retail.posterminal.master;

import java.util.ArrayList;
import java.util.List;

import org.codehaus.jettison.json.JSONException;
import org.codehaus.jettison.json.JSONObject;
import org.openbravo.model.pricing.pricelist.PriceList;
import org.openbravo.retail.config.OBRETCOProductList;
import org.openbravo.retail.posterminal.POSUtils;
import org.openbravo.retail.posterminal.ProcessHQLQuery;

public class Product extends ProcessHQLQuery {

  @Override
  protected List<String> getQuery(JSONObject jsonsent) throws JSONException {
    final OBRETCOProductList productList = POSUtils.getProductListByOrgId(jsonsent
        .getString("organization"));

    final PriceList priceList = POSUtils.getPriceListByOrgId(jsonsent.getString("organization"));

    if (productList == null) {
      throw new JSONException("Product list not found");
    }

    List<String> products = new ArrayList<String>();

    // regular products
    products
        .add("select pli.product.id as id, pli.product.searchKey as searchkey, pli.product.name as _identifier, pli.product.taxCategory.id as taxCategory, "
            + "pli.product.productCategory.id as productCategory, pli.product.obposScale as obposScale, pli.product.uOM.id as uOM, pli.product.uOM.symbol as uOMsymbol, pli.product.uPCEAN as uPCEAN, img.bindaryData as img "
            + ", pli.product.description as description "
            + ", pli.product.obposGroupedproduct as groupProduct "
            + ", pli.product.stocked as stocked "
            + ", pli.product.obposShowstock as showstock "
            + ", pli.bestseller as bestseller "
            + ", 'false' as ispack, "
            + "ppp.listPrice as listPrice, ppp.standardPrice as standardPrice, ppp.priceLimit as priceLimit, "
            + "ppp.cost as cost "
            + "FROM OBRETCO_Prol_Product as pli left outer join pli.product.image img, "
            + "PricingProductPrice ppp, "
            + "PricingPriceListVersion pplv "
            + "WHERE (pli.obretcoProductlist = '"
            + productList.getId()
            + "') "
            + "AND ("
            + "pplv.priceList.id = '"
            + priceList.getId()
            + "' AND "
            + "pplv.validFromDate = (select max(a.validFromDate) "
            + "  FROM PricingPriceListVersion a "
            + "  WHERE a.priceList.id = '"
            + priceList.getId()
            + "')"
            + ") AND ("
            + "ppp.priceListVersion.id = pplv.id"
            + ") AND ("
            + "pli.product.id = ppp.product.id"
            + ") AND "
            + "(pli.product.$incrementalUpdateCriteria) order by pli.product.name");

    // discounts which type is defined as category
    products
        .add("select p.id as id, p.name as searchkey, p.name as _identifier, p.discountType.id as productCategory, p.obdiscPrice as listPrice, p.obdiscPrice as standardPrice, p.obdiscUpc as uPCEAN, img.bindaryData as img, 'true' as ispack, 'false' as stocked"//
            + "  from PricingAdjustment as p left outer join p.obdiscImage img" //
            + " where p.discountType.obposIsCategory = true "//
            + "   and p.discountType.active = true " //
            + "   and p.active = true"//
            + "   and p.$readableClientCriteria"//
            + "   and (p.$incrementalUpdateCriteria) "//
        );

    return products;

  }
}

File: ${env.CATALINA_HOME}/webapps/openbravo/web/org.openbravo.retail.posterminal/js/model/product.js
/*
 ************************************************************************************
 * Copyright (C) 2012 Openbravo S.L.U.
 * Licensed under the Openbravo Commercial License version 1.0
 * You may obtain a copy of the License at http://www.openbravo.com/legal/obcl.html
 * or in the legal folder of this module distribution.
 ************************************************************************************
 */

/*global Backbone */

(function () {

  var Product = Backbone.Model.extend({
    modelName: 'Product',
    tableName: 'm_product',
    entityName: 'Product',
    source: 'org.openbravo.retail.posterminal.master.Product',
    dataLimit: 300,
    properties: ['id', 'searchkey', 'uPCEAN', 'uOM', 'uOMsymbol', 'productCategory', 'taxCategory', 'img', 'description', 'obposScale', 'groupProduct', 'stocked', 'showstock', 'bestseller', 'ispack', 'listPrice', 'standardPrice', 'priceLimit', 'cost', '_identifier', '_idx'],
    propertiesFilter: ['_identifier', 'searchkey', 'uPCEAN'],
    propertyMap: {
      'id': 'm_product_id',
      'uPCEAN': 'upc',
      'uOM': 'c_uom_id',
      'uOMsymbol': 'c_uom_symbol',
      'productCategory': 'm_product_category_id',
      'taxCategory': 'c_taxcategory_id',
      'img': 'img',
      'description': 'description',
      'obposScale': 'em_obpos_scale',
      'groupProduct': 'em_obpos_groupedproduct',
      'stocked': 'stocked',
      'showstock': 'em_obpos_showstock',
      'bestseller': 'bestseller',
      'ispack': 'ispack',
      'listPrice': 'listPrice',
      'standardPrice': 'standardPrice',
      'priceLimit': 'priceLimit',
      'cost': 'cost',
      '_identifier': '_identifier',
      '_filter': '_filter',
      '_idx': '_idx'
    },
    createStatement: 'CREATE TABLE IF NOT EXISTS m_product (m_product_id TEXT PRIMARY KEY , searchkey TEXT , upc TEXT, c_uom_id TEXT, c_uom_symbol TEXT, m_product_category_id TEXT, c_taxcategory_id TEXT, img TEXT, description TEXT, em_obpos_scale TEXT, em_obpos_groupedproduct TEXT, stocked TEXT, em_obpos_showstock TEXT, bestseller TEXT, ispack TEXT, listPrice NUMERIC, standardPrice NUMERIC, priceLimit NUMERIC, cost NUMERIC, _identifier TEXT, _filter TEXT, _idx NUMERIC)',
    dropStatement: 'DROP TABLE IF EXISTS m_product',
    insertStatement: 'INSERT INTO m_product(m_product_id, searchkey ,upc, c_uom_id, c_uom_symbol, m_product_category_id, c_taxcategory_id, img, description, em_obpos_scale, em_obpos_groupedproduct, stocked, em_obpos_showstock, bestseller, ispack, listPrice, standardPrice, priceLimit, cost, _identifier, _filter, _idx)  VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)',
    updateStatement: ''
  });

  var ProductList = Backbone.Collection.extend({
    model: Product
  });

  window.OB = window.OB || {};
  window.OB.Model = window.OB.Model || {};
  window.OB.Collection = window.OB.Collection || {};

  window.OB.Model.Product = Product;
  window.OB.Collection.ProductList = ProductList;
}());